### PR TITLE
TEP-0142: Introduce Params and Results into StepActions CRD

### PR DESCRIPTION
**Pull request by chitrangpatel**
This PR introduces `params` and `results` into the `StepAction` CRD. It allows the `StepAction` to declare the params it needs and the results it will produce.
The [follow up PRs](https://github.com/tektoncd/pipeline/issues/7259) will contain the interaction of how a `Task` referencing the `StepAction` resolves them.

All the tests were borrowed from `pkg/apis/pipeline/v1/...` that overlapped with `paramSpec` and `Results`.

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
[WIP] Introduces params and results into the StepAction CRD.
```
/kind feature

### CODE REVIEW

File: pkg/apis/pipeline/v1alpha1/param_types.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1alpha1/param_types.go:248]: This is not needed

File: pkg/apis/pipeline/v1alpha1/param_types.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1alpha1/param_types.go:175]: no need for this

File: pkg/apis/pipeline/v1alpha1/param_types.go
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1alpha1/param_types.go:248]: ah yes. good point. Copy and pasting issues 🤯

File: pkg/apis/pipeline/v1alpha1/param_types.go
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1alpha1/param_types.go:175]: Fixing. Copy and pasting 🤯

**Review by Yongxuanzhang — APPROVED** [review]

File: pkg/apis/pipeline/v1alpha1/param_types.go
Comment (jerop) [inline pkg/apis/pipeline/v1alpha1/param_types.go:1]: Can we not reuse v1.Params in v1alpha1.StepAction instead of creating v1alpha1.Params?

For example, in v1beta1.ResolutionRequest, we do use v1.Params: 
https://github.com/tektoncd/pipeline/blob/7069889e796b0b12e12a3d1095a5f1a3d98c2739/pkg/apis/resolution/v1beta1/resolution_request_types.go#L66

Trying to see if we can avoid this duplication

/hold

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1alpha1/param_types.go:1]: Good question. And great catch. I thought we needed it to be different because it was a different CRD under a different API version. Similar to how we have `v1beta1` and `v1` types for Tasks and Pipelines.

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1alpha1/param_types.go:1]: All of `StepActions`, `Tasks` and `Pipelines` are under `pkg/api/pipeline` while resolution is separate. May be codegen might have issues. 

But worth a try though.

**Comment by jerop** [comment]
@tektoncd/core-maintainers @tektoncd/core-collaborators please take a look 🙏🏾

File: pkg/apis/pipeline/v1alpha1/result_types.go
Comment (jerop) [inline pkg/apis/pipeline/v1alpha1/result_types.go:1]: how's this different from other results?

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1alpha1/result_types.go:1]: Right now, `StepActionResults` and `TaskResults` look the same. Once we add support for fetching Results from steps, `TaskResults` will look different from `StepActionResults`.

File: pkg/apis/pipeline/v1alpha1/param_types.go
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1alpha1/param_types.go:1]: It works.

File: pkg/apis/pipeline/v1alpha1/result_types.go
Comment (jerop) [reply pkg/apis/pipeline/v1alpha1/result_types.go:1]: Sounds great, thank you!

**Review by jerop — COMMENTED** [review]
/unhold
/lgtm

**Pull request merged**
